### PR TITLE
Improve implementation of "display-when" attribute

### PR DESCRIPTION
**Pull request by bertfrees**
@kalaspuffar @PaulRambags I've improved the implementation of the "display-when" attirbute:
- The issue of unused available space at the bottom of the page is fixed (see https://github.com/mtmse/dotify.library/issues/22).
- It is not required anymore to add a `keep="page"` on elements with `display-when`.
- In theory we could now support other conditions than `(! $starts-at-top-of-page)`.

### CODE REVIEW

**Review by kalaspuffar — APPROVED** [review]

File: src/org/daisy/dotify/formatter/impl/page/RowGroupProvider.java [hunk 1]
```diff
@@ -275,4 +276,24 @@ private boolean shouldAddGroupForEmptyContent() {
...
+
+    private List<RowImpl> makeInvisible(List<RowImpl> rows) {
+        if (invisible && rows.size() > 0) {
+            List<RowImpl> newRows = new ArrayList<>();
+            for (RowImpl r : rows) {
+                newRows.add(new RowImpl.Builder(r).invisible(true).build());
+            }
+            rows = newRows;
+        }
+        return rows;
```
Comment (PaulRambags) [inline src/org/daisy/dotify/formatter/impl/page/RowGroupProvider.java:297]: Code may be simplified - the whole body can be replace by
`return rows.stream().map(this::makeInvisible).collect(toList());`

**Review by PaulRambags — APPROVED** [review]

File: src/org/daisy/dotify/formatter/impl/page/RowGroupProvider.java (hunk 1 above)
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/RowGroupProvider.java:294]: Done. I chose to not replace the whole body though, for the sake of efficiency (not always needed to make a new list).

File: src/org/daisy/dotify/formatter/impl/page/RowGroupDataSource.java [hunk 2]
```diff
@@ -45,6 +46,7 @@
     private final RowGroupSequence data;
     private BlockContext bc;
     private Function<Integer, Integer> reservedWidths = x -> 0;
+	private static final Function<Integer, Boolean> topOfPage = x -> x == 0;
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/page/RowGroupDataSource.java]: Fixed.

File: src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java [hunk 3]
```diff
@@ -1133,6 +1133,8 @@ private BlockProperties blockBuilder(StartElement el) {
                 if (!att.getValue().replaceAll(" ", "")
                         .equalsIgnoreCase("(!$starts-at-top-of-page)")
```
[the remark this answers is not in the document]
Comment (PaulRambags) [reply src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:1134]: Why not leave a comment in the code that you could support other conditions as well?

**Review by PaulRambags — APPROVED** [review]

File: src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/obfl/ObflParserImpl.java:1134]: I updated the comment.

**Review by PaulRambags — APPROVED** [review]

**Review by kalaspuffar — CHANGES_REQUESTED** [review]
```
> Task :checkstyleMain
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:36: Wrong order for 'java.util.Iterator' import. [ImportOrder]
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:200: 'if' construct must use '{}'s. [NeedBraces]
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java:202: 'else' construct must use '{}'s. [NeedBraces]
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/RowGroupDataSource.java:19: Wrong order for 'java.util.List' import. [ImportOrder]
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/RowGroupDataSource.java:189: Line is longer than 120 characters (found 127). [LineLength]
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/RowGroupDataSource.java:272: TODO is preferred to FIXME.  e.g. "TODO: (ENG-123) -  Refactor when v2 is released." [RegexpSingleline]
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/RowGroupDataSource.java:273:51: 'typecast' is not followed by whitespace. [WhitespaceAfter]
[ant:checkstyle] [ERROR] /home/danielp/github/dotify.library/src/org/daisy/dotify/formatter/impl/page/RowGroupDataSource.java:274:60: 'typecast' is not followed by whitespace. [WhitespaceAfter]
```

**Comment by bertfrees** [comment]
Sorry, keep forgetting to run `gradle check`.

I fixed it and also rebased onto main.

**Review by kalaspuffar — APPROVED** [review]

**Comment by kalaspuffar** [comment]
PR approved, running regression.

**Pull request merged**